bcgov/GDX-Analytics-microservice
# Gdxdsd 7197 google search split

### This PR does the following:

1. Moves and/or copies the python scripts google_search.py, along with its associated config, ddl, and development files into a separate google_search folder
2. Creates new python virtual environment files that only use packages specific to google_search.py
3. Edits the README.md file to removed details specific to the google_mybusiness, and google_mybusiness_servicebc_derived microservices

Please note that some files have been modified for testing purposes. You can view the modified files in the ticket

### Login to the development environment:

1. Run the following command to login to EC2
```
$ awsmfa prod ######
Generating new IAM STS Token ...
STS Session Token generated and saved in profile mfa successfully.
$ microservice_ssm
```
2. Change the directory to the development branch:
```
$ cd /home/microservice/branch/GDXDSD-7197_google_search_split/google_search
```

### Review README.md
Review [README.md](https://github.com/bcgov/GDX-Analytics-microservice/blob/GDXDSD-7197_google_search_split/google_search/README.md) and check to makes sure there are no details specific to google_mybusiness and google_mybusiness_servicebc_derived. Also make sure the details about google_search.py makes sense. 
Please note that there are possibly still references to these other microservices where relevant. But these references shouldn't be related to how the microservice works or how the configs are structured

### Review moved files

Part of this changes includes moving files from the old combined folder into a separate one specific for this microservice. In this section, we will be reviewing to make sure all of these files were moved successfully
1. Review [google_search](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_google_search_split/google_search) to make sure the microservice scripts have been moved successfully
- google_search.py
2. Run the following command to get a list of the scheduled microservice's configs:
```
$ crontab -l | grep google_search
```
4. Review the files in [config.d](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_google_search_split/google_search/config.d) and make sure any config that appeared in the previous step is listed
5. Review the files in [dml](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_google_search_split/google_search/dml) and make sure any file listed in one of the previous step's configs has an associated file here.

### Review copied files

1. Review the files in [google_search](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_google_search_split/google_search) to make sure the following file exist
- .flake8
- pylintrc
- README.md

### Review created files
1. Review the files in [google_search](https://github.com/bcgov/GDX-Analytics-microservice/tree/GDXDSD-7197_google_search_split/google_search) to make sure the following file exist
- Pipfile
- Pipfile.lock

### Testing the microservice:
1. Run the following command to create the test tables:
```
$ microservice_rs
SET

snowplow=> CREATE TABLE test.gdxdsd_7197_googlesearch (LIKE google.googlesearch);
CREATE TABLE
snowplow=> GRANT SELECT ON TABLE test.gdxdsd_7197_googlesearch TO LOOKER;
GRANT
snowplow=> SELECT tablename FROM pg_tables WHERE tablename ILIKE '%7197%';
        tablename         
--------------------------
 gdxdsd_7197_googlesearch
(1 row)

snowplow=> \q
```
2. Run the following command to test the cmslite_user_data_to_redshift.py script and compare it to the expected output:
```
$ /home/microservice/.local/bin/pipenv run python3.8 google_search.py -o credentials_search.json -a credentials_search.dat -c config.d/GDXDSD-7197_config_search.json
Report: google_search.py

Config: config.d/GDXDSD-7197_config_search.json

Google Search derived table build not attempted
Microservice started at: 2024-11-06 13:54:20-0800 (PST), ended at: 2024-11-06 13:54:20-0800 (PST), elapsing: 0:00:00.000046.

Sites to process: 2
Successful API calls: 2
Failed API calls: 0
Failed loads to RedShift: 0


Objects loaded to S3 and copied to RedShift:
1: s3://sp-ca-bc-gov-131565110619-12-microservices/client/doug_test/GDXDSD-7197/google_gdx/googlesearch-search.onlinelearningbc.com-2024-11-01-2024-11-04.csv
2: s3://sp-ca-bc-gov-131565110619-12-microservices/client/doug_test/GDXDSD-7197/google_gdx/googlesearch-news.gov.bc.ca-2024-11-01-2024-11-04.csv
```
3. Run the following query to check if the data was loaded correctly:
```
$ microservice_rs
SET

snowplow=> SELECT count(*) from test.gdxdsd_7197_googlesearch;
 count  
--------
 195686
(1 row)

snowplow=> \q
```
4. Check the following files appear in S3 client: https://ca-central-1.console.aws.amazon.com/s3/buckets/sp-ca-bc-gov-131565110619-12-microservices?region=ca-central-1&bucketType=general&prefix=client/doug_test/GDXDSD-7197/google_gdx/&showversions=false
5. Run the following queries to drop the test tables:
```
$ microservice_rs
SET

snowplow=> drop table test.gdxdsd_7197_googlesearch;
DROP TABLE
snowplow=> \q
```
